File: Exemplos/App_eXperience_XamarinAndroid/Elgin_AppExperience_XamarinAndroid-font/XamarinAndroid-M8-omit_data/NFCE/It4r.cs
using System;
using System.Threading;
using Java.Util.Concurrent.Atomic;

using BR.Com.Daruma.Framework.Mobile;
using BR.Com.Daruma.Framework.Mobile.Exception;
using System.Text.RegularExpressions;
using Android.Util;

namespace M8
{
    public class It4r : It4rAbstract
    {
        //Valor utilizado para guardar o valor do tempo de emissão
        private static AtomicLong timeElapsedInLastEmission = new AtomicLong(0);
        //Obtejo utilizada para verificar se houve alguma exceção nas threads que executam as funções DarumaMobile
        private static volatile DarumaException capturedException = null;

        public It4r(DarumaMobile dmfObject) : base(dmfObject)
        { }

        public override void VenderItem(string descricao, string valor, string codigo)
        {
            Thread sellItemThread = new Thread(new ThreadStart(() => {
                try
                {
                    if (dmf.RCFVerificarStatus_NFCe() < 2)
                        dmf.ACFAbrir_NFCe("", "", "", "", "", "", "", "", "");

                    dmf.ACFConfICMS00_NFCe("0", "00", "3", "17.50");
                    dmf.ACFConfPisAliq_NFCe("01", "10.00");
                    dmf.ACFConfCofinsAliq_NFCe("01", "10.00");
                    dmf.ACFVenderCompleto_NFCe("17.50", "1.00", valor, "D$", "0.00", codigo, "21050010", "5102", "UN", descricao, "CEST=2300100;cEAN=SEM GTIN;cEANTrib=SEM GTIN;");
                }
                catch (DarumaException darumaException)
                {
                    darumaException.PrintStackTrace();
                    capturedException = darumaException;
                }
            }));

            sellItemThread.Start();

            try
            {
                sellItemThread.Join();
            }
            catch (Java.Lang.InterruptedException e)
            {
                e.PrintStackTrace();
            }

            if (capturedException != null)
                ThrowExceptionAndResetCapture();
        }

        public override void EncerrarVenda(string valorTotal, string formaPagamento)
        {
            Thread finishSaleThread = new Thread(new ThreadStart(() => {
                try
                {
                    if (float.Parse(valorTotal) < 0.01f)
                    {
                        throw new Java.Lang.IllegalArgumentException("Valor não aceito para o encerramento de venda!");
                    }
                    dmf.ACFTotalizar_NFCe("D$", "0.00");
                    dmf.ACFEfetuarPagamento_NFCe(formaPagamento, valorTotal);

                    timeElapsedInLastEmission = new AtomicLong(0);
                    long startTime = GetCurrentMillis();

                    dmf.TCFEncerrar_NFCe("ELGIN DEVELOPERS COMMUNITY");

                    long endTime = GetCurrentMillis();

                    timeElapsedInLastEmission = new AtomicLong(endTime - startTime);

                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\EstadoCFe", "0");

                    Log.Debug("MADARA", "NANI");
                }
                catch (DarumaException darumaException)
                {
                    darumaException.PrintStackTrace();
                    capturedException = darumaException;
                }
            }));

            finishSaleThread.Start();

            try
            {
                finishSaleThread.Join();
            }
            catch (Java.Lang.InterruptedException e)
            {
                e.PrintStackTrace();
            }

            if (capturedException != null)
                ThrowExceptionAndResetCapture();
        }

        public long GetCurrentMillis()
        {
            DateTime Jan1970 = new DateTime(1970, 1, 1, 0, 0, 0, DateTimeKind.Utc);
            TimeSpan javaSpan = DateTime.UtcNow - Jan1970;
            return (long)javaSpan.TotalMilliseconds;
        }

        public override void ConfigurarXmlNfce()
        {
            Thread configurateXmlNfceThread = new Thread(new ThreadStart(() => {
                try
                {
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\EmpPK", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\EmpCK", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("IDE\\cUF", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("EMIT\\CNPJ", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("EMIT\\IE", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("EMIT\\xNome", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("EMIT\\ENDEREMIT\\UF", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("EMIT\\CRT", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("EMIT\\CRT", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\TipoAmbiente", "<TIPO DE AMBIENTE>", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\Impressora", "EPSON", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\AvisoContingencia", "1", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\ImpressaoCompleta", "2", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\NumeracaoAutomatica", "1", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\HabilitarSAT", "0", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\IdToken", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\Token", "<INSERIR SEUS DADOS>", false);
                    dmf.RegAlterarValor_NFCe("IDE\\Serie", "133", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\NT\\VersaoNT", "400", false);
                    dmf.RegAlterarValor_NFCe("CONFIGURACAO\\EstadoCFe", "0", false);
                    dmf.RegPersistirXML_NFCe();

                    AdjustNfceNumber();
                }
                catch (DarumaException darumaException)
                {
                    darumaException.PrintStackTrace();
                    capturedException = darumaException;
                }
            }));

            configurateXmlNfceThread.Start();

            try
            {
                configurateXmlNfceThread.Join();
            }
            catch (Java.Lang.InterruptedException e)
            {
                e.PrintStackTrace();
            }

            if (capturedException != null)
                ThrowExceptionAndResetCapture();
        }

        private void ThrowExceptionAndResetCapture()
        {
            DarumaException lastException = capturedException;
            capturedException = null;
            throw lastException;
        }

        public string GetNumeroNota()
        {
            char[] numeroNota = new char[50];
            dmf.RInfoEstendida_NFCe("2", numeroNota);

            return new string(numeroNota).Trim();
        }

        public string GetNumeroSerie()
        {
            char[] serieNota = new char[50];
            dmf.RInfoEstendida_NFCe("5", serieNota);

            return new string(serieNota).Trim();
        }

        public AtomicLong GetTimeElapsedInLastEmission()
        {
            return timeElapsedInLastEmission;
        }

        //Função que busca a informação da nota mais alta pra série já escrita NFC-e para envio em cache, impedindo erro de duplicidade de NFC-e
        public void AdjustNfceNumber()
        {
            char[] retorno = new char[50];
            dmf.RRetornarInformacao_NFCe("NUM", "0", "0", "133", "", "9", retorno);
            string retornoAjustado = new string(retorno).Trim();
            string notaMaisAlta = Regex.Replace(retornoAjustado, "\\D+", "");
            int notaMaisAltaInt = int.Parse(notaMaisAlta) + 1;
            string proximaNota = notaMaisAltaInt.ToString();

            dmf.RegAlterarValor_NFCe("IDE\\nNF", proximaNota);
            dmf.RegPersistirXML_NFCe();
        }
    }
}

File: Exemplos/App_eXperience_FireMonkey/Elgin_AppExperience_DelphiFiremonkey-font/It4r.pas
unit It4r;

interface
  uses Elgin.JNI.Daruma,
//    System.Threading,
    System.Classes,
    System.SysUtils,
    FMX.Types,
    Androidapi.Helpers,
    Androidapi.JNIBridge,
    Androidapi.JNI.JavaTypes,
    System.DateUtils;
  type
    TIt4r = class
      private

      public
        function configurarXmlNfce():integer;
        function adjustNfceNumber():integer;
        function getNumeroNota(): string;
        function getNumeroSerie(): string;
        function venderItem(descricao, valor, codigo: string) : integer;
        function encerrarVenda(total, formaDePagamento: string): integer;
    constructor Create;
    end;

var
  Drm: TIt4r;
  timeElapsedInLastEmission: integer;
  dmfObject: TJDarumaMobile;
  dmf: JDarumaMobile;
  dDiff: Double;

implementation

{ TIt4r }


function TIt4r.adjustNfceNumber: integer;
var
  retorno: TJavaArray<Char>;
  retornoAjustado: JString;
  notaMaisAlta: JString;
  notaMaisAltaInt, resultt: integer;
  proximaNota: String;
begin
  retorno:= TJavaArray<Char>.Create(50);
  resultt:= dmf.rRetornarInformacao_NFCe(StringToJString('NUM'), StringToJString('0'),
                                StringToJString('0'),
                                StringToJString('133'),
                                StringToJString(''),
                                StringToJString('9'),
                                retorno);
  log.d('config - rRetornarInformacao_NFCe ~> ' + IntToStr(resultt));

  retornoAjustado := TJString.JavaClass.init(retorno);
  retornoAjustado:= retornoAjustado.trim();

  notaMaisAlta := retornoAjustado.replaceAll(StringToJString('\D+'), StringToJString(''));
  notaMaisAltaInt := StrToInt(JStringToString(notaMaisAlta)) + 1;
  proximaNota := (inttoStr(notaMaisAltaInt));

  resultt:= dmf.RegAlterarValor_NFCe(StringToJString('IDE\nNF'), StringToJString(proximaNota), False);
  log.d('config - nNF ~> ' + IntToStr(resultt));
  resultt:= dmf.RegPersistirXML_NFCe();
  log.d('config - Persiste_NumberNFCe ~> ' + IntToStr(resultt));

  Result:= resultt;

end;

function TIt4r.configurarXmlNfce: integer;
var
  resultt: integer;
  t: TThread;
begin
  t:= TThread.CreateAnonymousThread(procedure ()
  begin
    log.d('config - Resultados');
    resultt:= (dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\EmpPK'), StringToJString('<INSIRA SUA CHAVE AQUI>'), False));
    log.d('config - EmpPK ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\EmpCK'), StringToJString('<INSIRA SUA CHAVE AQUI>'), False);
    log.d('config - EmpCK ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('IDE\cUF'), StringToJString('43'), False);
    log.d('config - cUF ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('EMIT\CNPJ'), StringToJString('06354976000149'), False);
    log.d('config - CNPJ ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('EMIT\IE'), StringToJString('1470049241'), False);
    log.d('config - IE ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('EMIT\xNome'), StringToJString('ITFast'), False);
    log.d('config - xNome ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('EMIT\ENDEREMIT\UF'), StringToJString('RS'), False);
    log.d('config - ENDEREMIT\UF ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('EMIT\CRT'), StringToJString('3'), False);
    log.d('config - CRT ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\TipoAmbiente'), StringToJString('2'), False);
    log.d('config - TipoAmbiente ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\Impressora'), StringToJString('EPSON'), False);
    log.d('config - Impressora ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\AvisoContingencia'), StringToJString('1'), False);
    log.d('config - AvisoContingencia ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\ImpressaoCompleta'), StringToJString('2'), False);
    log.d('config - ImpressaoCompleta ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\NumeracaoAutomatica'), StringToJString('1'), False);
    log.d('config - NumeracaoAutomatica ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\HabilitarSAT'), StringToJString('0'), False);
    log.d('config - HabilitarSAT ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\IdToken'), StringToJString('000001'), False);
    log.d('config - IdToken ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\Token'), StringToJString('1A451E99-0FE0-4C13-B97E-67D698FFBC37'), False);
    log.d('config - Token ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('IDE\Serie'), StringToJString('133'), False);
    log.d('config - Serie ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\NT\VersaoNT'), StringToJString('400'), False);
    log.d('config - VersaoNT ~> ' + IntToStr(resultt));
    resultt:= dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\EstadoCFe'), StringToJString('0'), False);
    log.d('config - EstadoCFe ~> ' + IntToStr(resultt));
    resultt:= dmf.RegPersistirXML_NFCe();
    log.d('config - PersisteXML ~> ' + IntToStr(resultt));

    resultt:= adjustNfceNumber();
   end);

   t.Start;
   t.WaitFor;

   Result:= resultt;
end;

constructor TIt4r.Create();
begin
   dmf := TJDarumaMobile.JavaClass.inicializar(TAndroidHelper.Context,
              StringToJString('@FRAMEWORK(LOGMEMORIA=200;TRATAEXCECAO=FALSE;TIMEOUTWS=8000;SATNATIVO=FALSE);@SOCKET(HOST=192.168.210.94;PORT=9100;)'));
end;

function TIt4r.encerrarVenda(total, formaDePagamento: string): integer;
var
  resultt: integer;
  t: TThread;
  dBeforte, dAfter: TDateTime;
  dDiv: Double;
begin
  t:= TThread.CreateAnonymousThread(procedure ()
  begin
    log.d('venda - Encerrar');
    resultt:=  dmf.aCFTotalizar_NFCe(StringToJString('D$'), StringToJString('0.00'));
    log.d('venda - aCFTotalizar_NFCe ~> ' + IntToStr(resultt));
    resultt:=  dmf.aCFEfetuarPagamento_NFCe(StringToJString(formaDePagamento), StringToJString(total));
    log.d('venda - aCFEfetuarPagamento_NFCe ~> ' + IntToStr(resultt));
    dBeforte:= MilliSecondsBetween(Now, 0);
    resultt:=  dmf.tCFEncerrar_NFCe(StringToJString('ELGIN DEVELOPERS COMMUNITY'));
    dAfter:= MilliSecondsBetween(Now, 0);
    log.d('venda - tCFEncerrar_NFCe ~> ' + IntToStr(resultt));
    resultt:=  dmf.RegAlterarValor_NFCe(StringToJString('CONFIGURACAO\EstadoCFe'), StringToJString('0'));
    log.d('venda - RegAlterarValor_NFCe ~> ' + IntToStr(resultt));

    dDiv:= 1000;
    dDiff:= (dAfter - dBeforte) / dDiv;
  end);

  t.Start;
  t.WaitFor;

  Result:= resultt;
end;

function TIt4r.getNumeroNota: string;
var
  retorno: TJavaArray<Char>;
  res: JString;
  resultt: integer;
begin
  retorno:= TJavaArray<Char>.Create(50);
  resultt:= dmf.rInfoEstendida_NFCe(StringToJString('2'), retorno);
  log.d('venda - rInfoEstendida_NFCe ~> ' + IntToStr(resultt));

  res := TJString.JavaClass.init(retorno);
  result:= JStringToString(res.trim());
end;

function TIt4r.getNumeroSerie: string;
var
  retorno: TJavaArray<Char>;
  res: JString;
  resultt: integer;
begin
  retorno:= TJavaArray<Char>.Create(50);
  resultt:= dmf.rInfoEstendida_NFCe(StringToJString('5'), retorno);
  log.d('venda - rInfoEstendida_NFCe ~> ' + IntToStr(resultt));

  res := TJString.JavaClass.init(retorno);
  result:= JStringToString(res.trim());
end;

function TIt4r.venderItem(descricao, valor, codigo: string): integer;
var
  str_vazia:JString;
  resultt: integer;
  t: TThread;
begin
  t:= TThread.CreateAnonymousThread(procedure ()
  begin
    log.d('venda - Item');
    str_vazia:= StringToJString('');
    resultt:= dmf.rCFVerificarStatus_NFCe();
    log.d('venda - rCFVerificarStatus_NFCe ~> ' + IntToStr(resultt));
    if resultt < 2 then
    begin
      resultt:= dmf.aCFAbrir_NFCe(str_vazia, str_vazia,str_vazia, str_vazia, str_vazia, str_vazia, str_vazia, str_vazia, str_vazia);
      log.d('venda - aCFAbrir_NFCe ~> ' + IntToStr(resultt));
    end;

    resultt:= dmf.aCFConfICMS00_NFCe(StringToJString('0'), StringToJString('00'), StringToJString('3'), StringToJString('17.50'));
    log.d('venda - aCFConfICMS00_NFCe ~> ' + IntToStr(resultt));
    resultt:= dmf.aCFConfPisAliq_NFCe(StringToJString('01'), StringToJString('10.00'));
    log.d('venda - aCFConfPisAliq_NFCe ~> ' + IntToStr(resultt));
    resultt:= dmf.aCFConfCofinsAliq_NFCe(StringToJString('01'), StringToJString('10.00'));
    log.d('venda - aCFConfCofinsAliq_NFCe ~> ' + IntToStr(resultt));
    resultt:= dmf.aCFVenderCompleto_NFCe(StringToJString('17.50'),
                                     StringToJString('1.00'),
                                     StringToJString(valor),
                                     StringToJString('D$'),
                                     StringToJString('0.00'),
                                     StringToJString(codigo),
                                     StringToJString('21050010'),
                                     StringToJString('5102'),
                                     StringToJString('UN'),
                                     StringToJString(descricao),
                                     StringToJString('CEST=2300100;cEAN=SEM GTIN;cEANTrib=SEM GTIN;'));
    log.d('venda - aCFVenderCompleto_NFCe ~> ' + IntToStr(resultt));
//    resultt:= dmf.aCFTotalizar_NFCe(StringToJString('D$'), StringToJString('0.00'));
//    log.d('venda - aCFTotalizar_NFCe ~> ' + IntToStr(resultt));
//    resultt:= dmf.aCFEfetuarPagamento_NFCe(StringToJString('Dinheiro'), StringToJString(valor));
//    log.d('venda - aCFEfetuarPagamento_NFCe ~> ' + IntToStr(resultt));
  end);

  t.Start;
  t.WaitFor;

  Result:= resultt;
end;

initialization
  Drm:= TIt4r.Create;
end.
